crates/core/src/factories/core_fetch.rs

```diff
@@ -213,6 +214,17 @@ impl Fetch for CoreFetch {
         })
     }
 
+    fn notify_on_drained(&self, notify: futures::channel::oneshot::Sender<()>) {
+        let mut lock = self.state.lock().expect("poisoned");
+        if lock.requests.is_empty() {
+            if let Err(err) = notify.send(()) {
+                tracing::warn!(?err, "Failed to send notification on drained");
+            }
+        } else {
+            lock.drained_notify.push(notify);
+        }
+    }
```

Comment:
Could also do this with [tokio::sync::Notify](https://docs.rs/tokio/1.45.1/tokio/sync/struct.Notify.html#method.notify_waiters).

Reply:
Happy to make that change, I used `futures` because it's already in the API crate and Tokio isn't

Reply:
Ah, makes sense. I'm good either way.

Reply:
I think the suggestion would be better code, but the API crate is usable on its own for types and I like that it's quite small to build at the moment